pinot-core/src/main/java/org/apache/pinot/core/operator/filter/H3InclusionIndexFilterOperator.java

```diff
@@ -0,0 +1,160 @@
+import org.apache.pinot.segment.local.utils.H3Utils;
+import org.apache.pinot.segment.spi.IndexSegment;
+import org.apache.pinot.segment.spi.index.reader.H3IndexReader;
+import org.apache.pinot.spi.exception.BadQueryRequestException;
+import org.apache.pinot.spi.utils.BytesUtils;
+import org.locationtech.jts.geom.Geometry;
+import org.roaringbitmap.buffer.BufferFastAggregation;
+import org.roaringbitmap.buffer.ImmutableRoaringBitmap;
+import org.roaringbitmap.buffer.MutableRoaringBitmap;
+
+
+/**
+ * A filter operator that uses H3 index for geospatial data inclusion
+ */
+public class H3InclusionIndexFilterOperator extends BaseFilterOperator {
+
+  private static final String EXPLAIN_NAME = "INCLUSION_FILTER_H3_INDEX";
+
+  private static final String OPERATOR_NAME = "H3InclusionIndexFilterOperator";
+
+  private final IndexSegment _segment;
+  private final Predicate _predicate;
+  private final int _numDocs;
+  private final H3IndexReader _h3IndexReader;
+  private final LongSet _fullyCoverH3Cells;
+  private final LongSet _potentialCoverH3Cells;
+  private final boolean _isPositiveCheck;
+
+  public H3InclusionIndexFilterOperator(IndexSegment segment, Predicate predicate, int numDocs) {
+    _segment = segment;
+    _predicate = predicate;
+    _numDocs = numDocs;
+
+    List<ExpressionContext> arguments = predicate.getLhs().getFunction().getArguments();
+    Geometry geometry;
+    // Assume first argument is Literal, and second argument is IDENTIFIER for St_Contains.
+    assert arguments.get(1).getType() == Type.IDENTIFIER;
+    assert arguments.get(0).getType() == Type.LITERAL;
+    EqPredicate eqPredicate = (EqPredicate) predicate;
+    if (eqPredicate.getValue().equals("1")) {
+      _isPositiveCheck = true;
+    } else if (eqPredicate.getValue().equals("0")) {
+      _isPositiveCheck = false;
+    } else {
+      throw new BadQueryRequestException("Expected value for ST_Contain is 0 or 1");
+    }
+    // look up arg1's h3 indices
+    _h3IndexReader = segment.getDataSource(arguments.get(1).getIdentifier()).getH3Index();
+    // arg0 is the literal
+    geometry = GeometrySerializer.deserialize(BytesUtils.toBytes(arguments.get(0).getLiteral()));
+    // must be some h3 index
+    assert _h3IndexReader != null;
+
+    // get the set of H3 cells at the specified resolution which completely cover the input shape and potential cover.
+    Pair<LongSet, LongSet> fullCoverAndPotentialCoverCells =
+        H3Utils.coverGeometryInH3(geometry, _h3IndexReader.getH3IndexResolution().getLowestResolution());
+
+    _fullyCoverH3Cells = fullCoverAndPotentialCoverCells.getLeft();
+    _potentialCoverH3Cells = fullCoverAndPotentialCoverCells.getRight();
+  }
+
+  @Override
+  protected FilterBlock getNextBlock() {
+    // have list of h3 cell ids for polygon provided
+    // return filtered num_docs
+    ImmutableRoaringBitmap[] potentialMatchDocIds = new ImmutableRoaringBitmap[_potentialCoverH3Cells.size()];
+    int i = 0;
+
```

Comment:
(minor) Remove the empty lines in this method. I feel the empty lines are actually making the readability worse

Reply:
got it.